first attempt to implement matmul detection and optimization using th…

chelini opened the pull request:
…e matchers. Compared to the paper 'High-performace Generalized Tensor Operations: A compiler-Oriented Approach', the missing steps for detection are: check stride 1 for the induction variables i,j,k, check that the additional reads accesses that do not belong to matmul have stride 0 if the innermost loop is exchanged with any of the loops. Moreover, they also check that the last access is a write. Regrading optimization, the missing parts are: packing transformation and tile isolation. Moreover, I assume that the initial indexes are in the following order i-j-k with i outermost and manually set the tile sizes.

ftynse commented on islutils/pet_wrapper.cc:
```diff
@@ -167,4 +167,8 @@ pet_stmt *Scop::stmt(isl::id id) const {
   return nullptr;
 }
 
+IslCopyRefWrapper<isl::schedule> Scop::schedule() {
```
This looks like code I have written.  How about keeping it in a separate commit?  This makes clear who wrote what and hence who's to blame when it breaks.  It also avoids painful rebase management when I attempt to merge one of the branches where I do have a separate commit adding exactly the same code (44d11724a50ef669bc9e0c13aaf5615894b8c535 
63cb7b8 )

chelini replied:
Yes, it is your code. Can we integrate it into the master?

ftynse replied:
Just don't meld it into your commit,  [cherry-pick](https://git-scm.com/docs/git-cherry-pick) the original commits to your branch if you need them.

ftynse commented on test/CMakeLists.txt:
```diff
@@ -14,6 +14,7 @@ set(TEST_INPUTS
     strided_domain.c
     strided_domain_with_coefficients.c
     strided_domain_multi_dimensions.c
+    1mmWithOutInitStmt.c
```
Nit: "without" is a single word in English, there is no sense in capitalizing "out"

ftynse commented on test/inputs/1mmWithOutInitStmt.c:
```diff
@@ -0,0 +1,15 @@
+int tmp[1024][1024];
```
I would try with non-square matrices

ftynse commented on test/test_transformer.cc:
```diff
@@ -518,3 +518,211 @@ TEST(Transformer, Codegen) {
 }
+
+static isl::schedule_node tile(isl::schedule_node node,
+			       std::vector<int> tileSizes,
```
Please do _not_ mix tabs and spaces.  Generally, run clang-format with LLVM style on your code.

chelini replied:
Yeah sorry, I forgot to run clang-format.

pfaffe replied:
You can make clang-format into a git precommit hook, either to auto-format your commit or to reject it so that you can format it.

ftynse commented on test/test_transformer.cc:
```diff
@@ -518,3 +518,211 @@ TEST(Transformer, Codegen) {
+
+static isl::schedule_node tile(isl::schedule_node node,
+			       std::vector<int> tileSizes,
+  			       int defaultSize = 32) {
```
I find this interface quite confusing, I would have made it fail on an assertion (it's a test after all!) if the size of `tileSizes` mismatches the number of band members.  Try writing documentation for this function to understand my complaint :)

chelini replied:
Ok, I think I got your point. In this case, if I want to use assert in the method I should you cassert or the one provided by googletest? It is ok to mix both?

ftynse replied:
You cannot use those from googletest unless you are inside a function declared as test, so just use the C one.  Or have a manual check and call `ISLUTILS_DIE` from `die.h`.

ftynse commented on test/test_transformer.cc:
```diff
@@ -518,3 +518,211 @@ TEST(Transformer, Codegen) {
+  return node;
+}
+
+static isl::multi_union_pw_aff permuteDims(isl::multi_union_pw_aff ps, int firstDim,
```
maybe call this `swapDims` it does not support arbitrary permutations

ftynse commented on test/test_transformer.cc:
```diff
@@ -518,3 +518,211 @@ TEST(Transformer, Codegen) {
+  return ps;
+}
+
+TEST(Transformer, matmul) {
```
Please capitalize the name of the test, it may also be helpful to make it more self-descriptive like `MatchMatmul`

ftynse commented on test/test_transformer.cc:
```diff
@@ -518,3 +518,211 @@ TEST(Transformer, Codegen) {
+    [&node] (isl::schedule_node n) {
+      if(isl_schedule_node_band_n_member(n.get()) < 3)
+        return false;
+      else { node = n; return true; }
```
Either use braces for both the "if" and "else" parts, or don't use them at all.  Also consider starting each statement from a new line of you have more than one and use braces.

ftynse commented on test/test_transformer.cc:
```diff
@@ -518,3 +518,211 @@ TEST(Transformer, Codegen) {
+  // clang-format on
+
+  bool matched = false;
+  if(ScheduleNodeMatcher::isMatching(matcher, root.child(0))) {
```
This control flow is useless.  Just do `auto matched = ScheduleNodeMatcher::isMatching(matcher, root.child(0));`

ftynse commented on test/test_transformer.cc:
```diff
@@ -518,3 +518,211 @@ TEST(Transformer, Codegen) {
+		      access(_C, _k, _j));
+  auto matchedReads = 
+    match(reads, psRead);
+  //std::cout << "matched Reads " << matchedReads.size() << std::endl;
```
Let's not commit commented-out code if possible.

ftynse commented on test/test_transformer.cc:
```diff
@@ -518,3 +518,211 @@ TEST(Transformer, Codegen) {
+  // this inspecting the placeholder for the write and the read.
+  // They should be equal. 
+ 
+  matched = matched && (matchedWrites.size() == 1);
```
Why not just `ASSERT_EQ(matchedWrites.size(), 1u)` ?  It would fail the test early without executing anything else.

ftynse commented on test/test_transformer.cc:
```diff
@@ -518,3 +518,211 @@ TEST(Transformer, Codegen) {
+  // They should be equal. 
+ 
+  matched = matched && (matchedWrites.size() == 1);
+  matched = matched && (matchedReads.size() == 1);
```
The name `matchedReads` makes it slightly confusing.  It implies it contains the matched read relations.  I expect three of those, but you check there is only one.  (I know the variable contains _match descriptors_, indexed by placeholders.)  Maybe `readMatches` is a better name here.

ftynse commented on test/test_transformer.cc:
```diff
@@ -518,3 +518,211 @@ TEST(Transformer, Codegen) {
+
+  // tmp[j][i] = alpha * A[i][k] * B[k][j] + tmp[i][j]
+  // pass the checks. this is because we do not link
+  // read and write at the moment. Placeholder are _not_
```
Indeed.  Or figure out a way to reuse placeholders between different calls.  Or take a union of reads and writes and differentiate them by tags (may need a different pattern/candidate class than we already have).

ftynse commented on test/test_transformer.cc:
```diff
@@ -518,3 +518,211 @@ TEST(Transformer, Codegen) {
+  // step 1. (not implemented, just assumed)
+  // we assume the indexes are already in
+  // the "good" order. (i,j,k).
+  // Line 1285 to 1291 in ScheduleOptimizer.cpp
```
ScheduleOptimizer.cpp is not in this codebase, so I wouldn't base the documentation based on it.  Especially line numbers that are prone to changes without notice :)

Also, it should be easy to check the right order, you have access to `inputDimPos`...

chelini replied:
Got your point I will do it.

ftynse commented on test/test_transformer.cc:
```diff
@@ -518,3 +518,211 @@ TEST(Transformer, Codegen) {
+  // set tile values manually
+  int dimOutNum = isl_schedule_node_band_n_member(node.get());
+  std::vector<int> tileSizes(dimOutNum);
+  tileSizes[dimOutNum - 3] = /*96*/32;
```
This contradicts the comment above.  And this is exactly the reason of me insisting on not committing commented-out code.

ftynse commented on test/test_transformer.cc:
```diff
@@ -518,3 +518,211 @@ TEST(Transformer, Codegen) {
+  isl::schedule_node tiledNode = tile(node, tileSizes);
+  auto tileSchedule = tiledNode.band_get_partial_schedule();
+  auto pointSchedule = tiledNode.child(0).band_get_partial_schedule();
+  // id for mark nodes.
```
mark nodes are useless, polly does them for annotation, but we don't need them in the test IMO

chelini replied:
Ok I can remove them I think with mark not it is easy to check what's going on while you are transforming.

ftynse replied:
Keep them if you want but use them in checks then :)

ftynse commented on test/test_transformer.cc:
```diff
@@ -518,3 +518,211 @@ TEST(Transformer, Codegen) {
+          newPartialSchedule = permuteDims(newPartialSchedule, dimOutNum - 3, dimOutNum - 1);
+          return newPartialSchedule;},
+        mark(idPoint,
+	  band(pointSchedule))));
```
one more indent here

chelini replied:
mm, I saw it differently maybe because I am mixing tab with space.

ftynse commented on test/test_transformer.cc:
```diff
@@ -518,3 +518,211 @@ TEST(Transformer, Codegen) {
+        [&node, &tileSchedule, &dimOutNum]() {
+          auto newPartialSchedule = permuteDims(tileSchedule, dimOutNum - 2, dimOutNum - 1);
+          newPartialSchedule = permuteDims(newPartialSchedule, dimOutNum - 3, dimOutNum - 1);
+          return newPartialSchedule;},
```
please put the closing brace on a new line

ftynse commented on test/test_transformer.cc:
```diff
@@ -518,3 +518,211 @@ TEST(Transformer, Codegen) {
+  // match micro-kernel
+  auto matcherMicroKernel = [&]() {
+    using namespace matchers;
+    // clang-format off
```
there's no need to turn the formatter of if you don't have the tree formatting in can break :)

ftynse commented on test/test_transformer.cc:
```diff
@@ -518,3 +518,211 @@ TEST(Transformer, Codegen) {
+  // For the micro and macro kernels we assume
+  // given values for the tile size.
+  using namespace builders;
+  // tiling - Mc=96, Nc=2048, Kc=512
```
What are `Mc`, `Nc` and `Kc` ?   Your input code has fixed sizes...

chelini replied:
These are tile size values. The values are obtained using an analytical model that makes use of TargetTransform pass in LLVM. I basically got the values for my machine and check that I am generating the same code as Polly does.

ftynse replied:
Ok, just explain what `Mc` and `Nc` etc. mean and where they should go

ftynse commented on test/test_transformer.cc:
```diff
@@ -518,3 +518,211 @@ TEST(Transformer, Codegen) {
+  EXPECT_TRUE(matched);
+  
+  // create micro-kernel
+  // tiling Mr = 4 and Nr = 8 and 1.
```
What are `Mr` and `Nr` ?

ftynse commented on test/test_transformer.cc:
```diff
@@ -518,3 +518,211 @@ TEST(Transformer, Codegen) {
+    //std::cout << root.to_str() << std::endl;
+  }
+/*
+  // isolate full-tile from partial tile
```
let's node commit commented-out code.

ftynse commented on test/test_transformer.cc:
```diff
@@ -518,3 +518,211 @@ TEST(Transformer, Codegen) {
+*/    
+  std::cout << petScop.codegen() << std::endl;
+  petScop.schedule() = root.get_schedule();
+  std::cout << petScop.codegen() << std::endl;
```
How about you check for _some_ pattern in the code instead of just dumping it.  Like the new loops are indeed produced or something?

ftynse commented on test/test_transformer.cc:
```diff
@@ -518,3 +518,211 @@ TEST(Transformer, Codegen) {
+  tileSizes[dimOutNum - 2] = /*2048*/32;
+  tileSizes[dimOutNum - 1] = /*512*/32;
+  // tile node and get partial schedule 
+  isl::schedule_node tiledNode = tile(node, tileSizes);
```
I don't quite get this.  You transform the node using isl's function, then essentially drop the transformed node and transform the original node _again_ with builders.  What's the point?  Maybe just extract the "tile" function that computes a pair of schedules (tiles, points) given the original schedule and sizes?

chelini replied:
Yes, it is a bit silly. I just want to get the new schedule for the tile and the point tile. To do so I transform the node and get the new partial schedules and re-build it. Do you have suggestions to improve it? What do you mean by "extracting the tile function"?

ftynse replied:
You don't need to transform the node, you just need the two `multi_union_pw_aff`s for the parent and the child node.  There may be a function that does that in isl, or it's quite easy to implement one (divide by multi-val, or take modulo by multival).

ftynse reviewed (commented):
I'm a little nitpicky on this one, but seems to go in the right direction.  I wish we did it earlier.

ftynse reviewed (commented):
Thanks.  Please squash all "fix" commits into the first one and split its description into a short part (76 characters) and a longer part (everything else separated by a newline from the first 76 characters); and it will be ready to go.

chelini committed:
First attempt at implementing Polly matmul optimiztion.

The optimization performed is based on the paper: High-Performance Generalized Tensor Operations: A Compiler-Oriented Approach. The algorithm implemented is at page 7 of the mentioned paper. The missing transformation is the packing of arrays A and B. For tile size we assume manually set values, in this particular case (32, 32, 32) for the macro-kernel and (2, 2, 1) for the micro-kernel. Notice that there is a mismatch between what Polly does and the optimization steps described in the paper. This should be clarified.

chelini commented:
I have squashed the commits and push back. Please let me know if it looks good.

ftynse commented:
Rebased and pushed to master as 6761257 

I forgot that I asked you to separate out the pre-existing code before asking you to squash everything, so I did that myself by merging #9 and rebasing.  

Generally, please consider making your git commits as unitary logical changes, not just snapshots of your working directory at some given time.  This helps tracking down problematic changes or reverting them if necessary.  That's why I ask to squash in review changes, and separate out changes that were made elsewhere.

chelini commented:
Thanks Alex.
